pytorch_toolkit/formula_recognition/im2latex/models/text_recognition_heads/attention_based.py

```diff
@@ -0,0 +1,280 @@
+        logits = torch.stack(logits, dim=1)  # [B, MAX_LEN, out_size]
+        targets = torch.max(torch.log(logits).data, dim=2)[1]
+        return logits, targets
+
+    def decode_without_formulas(self, h, c, output, row_enc_out, b_size, device):
+        logits = []
+        logit = None
+        for _ in range(self.max_len):
+            if logit is not None:
+                tgt = torch.reshape(torch.max(logit.data, dim=1)[
+                                    1], (b_size, 1)).clone().detach()
+            else:
+                tgt = torch.tensor([[START_TOKEN]] * b_size)
+            tgt = tgt.to(device)
+            # one step decoding
+            h, c, output, logit = self.step_decoding(
+                h, c, output, row_enc_out, tgt)
+
+            logits.append(logit)
+        logits = torch.stack(logits, dim=1)  # [B, MAX_LEN, out_size]
+        targets = torch.max(torch.log(logits).data, dim=2)[1]
+        return logits, targets
+
+    def decode_with_bs(self, h, c, O_t, row_enc_out, b_size, device):
+        max_len = self.max_len
+        # initializing empty list of candidates
+        # at first we have only one candidate (START_TOKEN) per image with score = 1
+        candidates_list = []
+        candidates_list.append(Candidate(
+            score=0.,
+            dec_state=(h, c),
+            Out_t=O_t,
+            targets=[START_TOKEN],
+            logits=[]
+        ))
+        # the first prediction is empty with score = 1
+        logits = [[]] * b_size
+        targets = [[]] * b_size
+
+        for _ in range(max_len):
+            # pushing every candidate into step decoding to get probable tracks
+            all_candidates = self.decode_with_bs_step(
+                row_enc_out, candidates_list, device)
+            all_candidates.sort(key=lambda cand: cand.score)
+            candidates_list = all_candidates[:self.beam_width]
+
+        logits = candidates_list[0].logits
+        targets = candidates_list[0].targets[1:]
+        logits = torch.stack(logits, dim=0)
+        logits = logits.unsqueeze(dim=0)
+        logits = logits.to(device)
+        targets = torch.tensor(targets).to(device)
+        targets = targets.unsqueeze(dim=0)
+
+        return logits, targets
+
+    def encode(self, encoded_imgs):
+        encoded_imgs = encoded_imgs.permute(0, 2, 3, 1)  # [B, H', W', LSTM_INP_CHANNELS]
+        # Prepare data for Row Encoder
+        # poccess data like a new big batch
+        B, H, W, out_channels = encoded_imgs.size()
+
+        encoded_imgs = encoded_imgs.contiguous().view(B*H, W, out_channels)
+
+        # prepare init hidden for each row
+        init_hidden_h = self.V_h_0.unsqueeze(
+            1).expand(self.V_h_0.shape[0], B*H, self.V_h_0.shape[1]).contiguous()
+        init_hidden_c = self.V_c_0.unsqueeze(
+            1).expand(self.V_h_0.shape[0], B*H, self.V_h_0.shape[1]).contiguous()
+        init_hidden = (init_hidden_h, init_hidden_c)
+
+        # Row Encoder
```

Comment:
Where do this idea come from?

Reply:
1. Did you try without encoding?
2. Did you try column encoding?
3. Did you try simultaneous row and column encoding?

Reply:
According to the [original paper](https://arxiv.org/pdf/1609.04938.pdf) (section "Row encoder"): 
"In this model, the new feature grid V is created from V˜ by running an RNN across each row of that input. Recursively for all rows h ∈ {1, . . . , H} and columns w ∈ {1, . . . , W}, the new features are defined as Vhw = RNN(Vh,w−1, Vhw). In order to capture the sequential order information in vertical direction, we use a trainable initial hidden state Vh,0 for each row, which we refer to as positional embeddings."